File: javascript/saddle-points/saddle-points.js
export class Matrix {
  constructor(str) {
    this.matrix = str.split('\n').map(r => r.trim().split(' ').map(Number));
    let cols = [];
    for (let i = 0; i < this.matrix[0].length; i++) {
      let col = [];
      for (let j = 0; j < this.matrix.length; j++) {
        col.push(this.matrix[j][i]);
      }
      cols.push(col);
    }
    this.cols = cols;
  }

  get rows() {
    return this.matrix;
  }

  get columns() {
    return this.cols;
  }

  get saddlePoints() {
    let saddlePoints = [];
    for (let r = 0; r < this.rows.length; r++) {
      for (let c = 0; c < this.columns.length; c++) {
        if (this.matrix[r][c] === Math.max(...this.matrix[r]) && 
        this.matrix[r][c] === Math.min(...this.columns[c])) saddlePoints.push([r,c]);
      }
    }
    return saddlePoints;
  }
}

File: python/saddle-points/saddle_points.py
def saddle_points(matrix):
    if not matrix:
        return []
    row_lens = [len(row) for row in matrix]
    if max(row_lens) != min(row_lens):
        raise ValueError('Invalid matrix')
    points = []
    row_max = [max(row) for row in matrix]
    col_min = [min(col) for col in [[row[i] for row in matrix] for i in range(len(matrix[0]))]]
    for row in range(len(matrix)):
        for col in range(len(matrix[0])):
            value = matrix[row][col]
            if value == row_max[row] and value == col_min[col]:
                points.append({"row": row + 1, "column": col + 1})
    return points
